core/store/hashtree/mem/tree.go

```diff
@@ -0,0 +1,477 @@
+		if n.GetType() == leafNodeType {
+			counter++
+		}
+	})
+
+	return counter
+}
+
+// Search returns the value associated to the key if it exists, otherwise nil.
+func (t *Tree) Search(key []byte, path *Path) ([]byte, error) {
+	if len(key) > t.maxDepth {
+		return nil, xerrors.Errorf("mismatch key length %d > %d", len(key), t.maxDepth)
+	}
+
+	// Build the big int representation of the key that is used for bitwise
+	// operations.
+	bi := new(big.Int)
+	bi.SetBytes(key)
+
+	value := t.root.Search(bi, path)
+
+	if path != nil {
+		path.root = t.root.GetHash()
+	}
+
+	return value, nil
+}
+
+// Insert inserts the key in the tree.
+func (t *Tree) Insert(key []byte, value []byte) error {
+	if len(key) > t.maxDepth {
+		return xerrors.Errorf("mismatch key length %d > %d", len(key), t.maxDepth)
+	}
+
+	// Build the big int representation of the key that is used for bitwise
+	// operations.
+	bi := new(big.Int)
+	bi.SetBytes(key)
+
+	t.root = t.root.Insert(key, value, bi)
+
+	return nil
+}
+
+// Delete removes a key from the tree.
+func (t *Tree) Delete(key []byte) error {
+	if len(key) > t.maxDepth {
+		return xerrors.Errorf("mismatch key length %d > %d", len(key), t.maxDepth)
+	}
+
+	bi := new(big.Int)
+	bi.SetBytes(key)
+
+	t.root = t.root.Delete(bi)
+
+	return nil
+}
+
+// Update updates the hashes of the tree.
+func (t *Tree) Update(fac crypto.HashFactory) error {
+	prefix := new(big.Int)
+
+	_, err := t.root.Prepare(t.nonce[:], prefix, fac)
+	if err != nil {
+		return xerrors.Errorf("failed to prepare: %v", err)
+	}
+
+	return nil
+}
+
+// Clone returns a deep copy of the tree.
+func (t *Tree) Clone() *Tree {
+	return &Tree{
+		nonce:    t.nonce,
+		maxDepth: t.maxDepth,
+		root:     t.root.Clone(),
+	}
+}
+
+// EmptyNode is leaf node with no value.
+type EmptyNode struct {
+	depth uint16
+	hash  []byte
+}
+
+// NewEmptyNode creates a new empty node.
+func NewEmptyNode(depth uint16) *EmptyNode {
+	return &EmptyNode{
+		depth: depth,
+	}
+}
+
+// GetHash implements mem.TreeNode. It returns the hash of the node.
+func (n *EmptyNode) GetHash() []byte {
+	return append([]byte{}, n.hash...)
+}
+
+// GetType implements mem.TreeNode. It returns the empty node type.
+func (n *EmptyNode) GetType() byte {
+	return emptyNodeType
+}
+
+// Search implements mem.TreeNode. It always return a empty value.
+func (n *EmptyNode) Search(key *big.Int, path *Path) []byte {
+	if path != nil {
+		path.leaf = n
+	}
+
+	return nil
+}
+
+// Insert implements mem.TreeNode. It replaces the empty node by a leaf node
+// that contains the key and the value.
+func (n *EmptyNode) Insert(key, value []byte, bi *big.Int) TreeNode {
+	return NewLeafNode(n.depth, key, value)
+}
+
+// Delete implements mem.TreeNode. It ignores the delete as an empty node
+// already means the key is missing.
+func (n *EmptyNode) Delete(key *big.Int) TreeNode {
+	return n
+}
+
+// Prepare implements mem.TreeNode. It updates the hash of the node and return
+// the digest.
```

Comment:
```suggestion
// Prepare implements mem.TreeNode. It updates the hash of the node and return
// the digest.
//
// We build and hash the following datagram:
//
// | TYPE | NONCE | PREFIX | DEPTH |
```

Reply:
Similarly to the other comment, the paper explains in details how the hash is created, and *why*. A pointer to the paper should be sufficient ?